### DISCUSSION TITLE
Look Modification Tools

### DISCUSSION THREAD

**Discussion by antonmeleshkevich** [jedypod/open-display-transform#10] [General]
Hi Jed!
Is this ok that I created the discussion here? Not sure what is the best place for this. I've never used github before, so I don't know if the discussions section here is the right place for this or not.

I've checked out your new look tools and they are awesome! Thanks for sharing them! Is there a chance they all will be combined in a one big DCTL? So it would be possible to change defaults in notepad and to use it as a DCTL LMT without loading it into DCTL OFX. It would also be helpful if there will be a switch for changing input primaries between AP0 and AP1, so it could be used as proper LMT.

Also I found one thing with NotoriousSix_Value and if you don't mind, I'd like to inform you about it, just in case. I'm talking about these dark lines.
![Screenshot 2021-10-18 061200](https://user-images.githubusercontent.com/34407001/137650410-9c87bbe1-32a9-4c0d-8491-220ff0557eed.jpg)

Also there are sharp transitions between primary colors with NotoriousSix_Vibrance.
![Screenshot 2021-10-18 062418](https://user-images.githubusercontent.com/34407001/137651776-83bf3bcb-47ee-4a48-b8eb-c82434d4a8bf.jpg)

They create artifacts with real world footage that has transitions from blue to red colors. This is not specific to NotoriousSix_Vibrance though, I get the same sharp transitions with all the tools for "subtractive" saturation I've ever seen (they are all based on HSV). But since you make all these wonderful tools I thought you probably know how to make it more clean in those transition regions.

And I really like how NotoriousSix_Vibrance and  NotoriousSix_Value do not produce any out of gamut colors if the source doesn't have negative values. All the similar tools I've seen Immediately go below 0.
And while I was writing this I checked NotoriousSix_HueShift and found that it produce some values out of 0-1 range.

Also I would like to clarify that your tools are amazing! The only reason why I report so many little things is that I really like them.

And I have one thing I would also like to report about OpenDRT and to demonstrate on a real footage that it doesn't preserve some colors hue (not LEDs and neons, just a red jacket) between different HDR Output Transforms. Where could I do that? And should I at all?

**Comment by jedypod** [comment]
Hey Anton! Good idea starting a discussion thread, I like it :smile: 

Been meaning to reply for a while, but have been pretty busy, sorry about the delay.

> one big DCTL?

Yes I have been thinking about making a library for shared code for the look tools also. If the functions for the look tools live in that library also, the look tools themselves could be pretty light. And a "preset" look (a look tool with a fixed set of parameters" could be a DCTL of just a few lines. It could be pretty slick actually, I think it's a good idea. I will implement these changes once I iron out a few more changes and the tools are a bit more stable.

>  sharp transitions between primary colors

Yes, most of the look tools use a linear extraction of hue. This creates a cusp at the boundaries between the primary and secondary hue angles. Those cusps could easily be smoothstepped though if it is a problem. I could add an optional checkbox pretty easily. I'll do some more testing to see if this would break anything unexpected.

> And while I was writing this I checked NotoriousSix_HueShift and found that it produce some values out of 0-1 range.

Yes I know. It will push some hues out of gamut. This can be okay if you are dealing with reflected surface colors. The appearance of the shifted hue can be natural sometimes. But sometimes it does not look natural. I was thinking of adding a limit slider or something like this to control how far colors are allowed to be pushed out of gamut. I will experiment with this and see if it works.

> And I have one thing I would also like to report about OpenDRT and to demonstrate on a real footage that it doesn't preserve some colors hue (not LEDs and neons, just a red jacket) between different HDR Output Transforms. Where could I do that? And should I at all?

I would be curious to see exact settings you are using, test images, and a description of this if you are up for it. Feel free to make an issue and we can talk about it.

Thanks for the useful feedback!

**Reply by antonmeleshkevich** [reply]
> This creates a cusp at the boundaries between the primary and secondary hue angles.

At first I've showed HSY color model results here, but later I found it looks nice with synthetic images only, but introduces artifacts into the real footage, so I deleted everything I've written in this post before.

**Comment by antonmeleshkevich** [comment]
I really like how I can get a nice starting point LMT with Vibrance tool, especially when applied to darker parts of the image only, but I found chromaticity-linear option very noisy with low saturated objects like18% gray card for example. Is there a way to tweak its behavior with these colors? The first thing that comes to my mind is to gradually exclude low saturated pixels from the effect. But it probably will also exclude a lot of skin tone colors. So probably there is a better way for this. Or probably there is some color model that would work good for selecting saturation for this.
And is it possible (without rewriting everything of course) to change chromaticity-linear checkbox with a slider that blends between two options? I found chromaticity-linear making darker skin tones too yellow, and with chromaticity-linear turned off they are too reddish. So blending between the two would help.

**Reply by jedypod** [reply]
Hey Anton! 
Thanks a lot for the suggestions. It is very helpful.

I went back and did a bunch more work on the vibrance tool. I came up with something a bit more sensible I think:
https://github.com/jedypod/open-display-transform/commit/40f76d3a49b141eb77fb340e928c3cb8b6bef5f1

I realized (after many hours of confused color pragmatist poking around), that I had neglected to multiply the hue angle extractions by chroma, which is what was causing the unacceptable amount of noise near grey values in the image. Changing this greatly improves the results I think. I also went in and did some tweaks and improvements to the parameterization. I like your idea of a slider going between hue-linear and per-component style non-linear. I've added this "hue-linear" slider, and changed the parameterization a bit so that the results between the two approaches are similar. I think it works pretty well but I'm curious to hear what you think.

Have a play and let me know :)

**Reply by antonmeleshkevich** [reply]
The noise level is much better now, thank you! It is still a bit more noisy compared to hue-linear slider to zero. And they both introduce a bit more noise compared to a regular saturation knob. But I guess it is impossible to achieve the same level of noise with this kind of correction. Also I tried to use Resolve qualifier and exclude low saturated pixels. Of course it helped with the noise, but it also excluded darkest skin tones from the effect. So maybe it is about finding the good balance. And the darkest skin tones is actually the main thing that I found Vibrance tool useful for :) I mean, it's also great for other things, but it is the only tool I found suitable for dealing with low saturated skin tone shadows after OpenDRT. So I have no idea if saturation qualifier is good for this or not. Just my thoughts after using this tool. Just in case, maybe you'll find it useful in some way :)
And regarding desaturated skin tone shadows - what tool is most suitable for compensating this? Is it Vibrance or ZoneSaturation? I didn't test ZoneSaturation for adding saturation into the shadows, to be honest. I only used it for de-saturating the highlights.

Thanks a lot for adding hue-linear slider!

And I would also like to show an example where those sharp transitions from one hue to another, I mentioned above, introduce artifacts.
This image is from a music video I've been working with a few months ago. It was shot on RED Comodo, but somebody did something wrong during shooting, so it was recorded as a Rec709(-ish) video, not a scene referred log or RAW. And I had to crop it as this is the only way I can post this. I know, this is now the best example, but this is the only one I have at this moment.

So here is the source image ( but it has de-noiser baked in, I only have this as a de-noised DPX now). It is rec709-ish image from camera. It is a stained glass in the church out of focus.

![1 48 1_1 48 1](https://user-images.githubusercontent.com/34407001/143725622-149eb959-0139-44f2-b1e3-2575ef091951.jpg)

Then I converted it to ACEScg using  OpenDRT 1886 inverse. Then I added some contrast with gamma applied to this linear encoded image.
(Probably I should've use offset to lower the shadows only, because the image has lifted shadows because of lens flares)
Then it is rendered to display with OpenDRT again.

![1 48 2_1 48 2](https://user-images.githubusercontent.com/34407001/143725715-95863f8d-fa88-4e9f-afa6-d8cc2b98ee98.jpg)

Here I applied Vibrance tool with Global slider to maximum, hue-linear to 0 and zone extract turned off.
![hue-lin-0_1 48 3](https://user-images.githubusercontent.com/34407001/143725749-90ab1690-036c-4be2-9970-6590e5c5ae72.jpg)

The same but with hue-linear to 0
![hue-lin-1_1 48 4](https://user-images.githubusercontent.com/34407001/143725783-6bd2e503-1ac1-485f-b025-5b3e13224a23.jpg)

And here is Value tool with all sliders to the max (except for Custom). This is Value tool only, without Vibrance.
![value_1 48 5](https://user-images.githubusercontent.com/34407001/143725877-c3fe18e6-5cf3-42f4-8e15-b8720c908edb.jpg)

I totally agree that this is a nitpicking. I just wanted to show this effect not just with the test image. And this is probably not that matter, as rarely will anyone use the tools with everything to the max.

Oh, and I forgot about dpx https://drive.google.com/file/d/16MoI9HSz6-9-uue-Kt8aDc9SvmHbqIVk/view

**Comment by ValeoMooha** [comment]
Hello Jed. Thank you for your tools!
Out of all, only GamutConvert, LogConvert and LinealGrade works on Mac (Ventura 13.1, Metal\OpenCL, Davinci 18.1)
Can you suggest some fix? Thank you.

**Reply by jedypod** [reply]
Hey @ValeoMooha - sorry just seeing this post now.

I mainly dev on my linux workstation, but I broke out my old macbook and tested all dctls (resolve 18.1.2, macos 10.15.7) and they all work. 

Hard for me to know what's going on without a reproducible test case... Can you paste the log in `/Library/Application Support/Blackmagic Design/DaVinci Resolve/logs/davinci_resolve.log` ? It might give us some clues.
